Meadowlark-js/packages/meadowlark-core/src/plugin/PluginLoader.ts

```diff
@@ -2,35 +2,87 @@
 // See the LICENSE and NOTICES files in the project root for more information.
-/* eslint-disable global-require */
+
 import dotenv from 'dotenv';
-import { Logger } from '../helpers/Logger';
-import { MeadowlarkBackendPlugin } from './backend/MeadowlarkBackendPlugin';
-import { NoMeadowlarkBackendPlugin } from './backend/NoMeadowlarkBackendPlugin';
+import { Logger } from '../Logger';
+import { DocumentStoreInitializer } from './backend/DocumentStoreInitializer';
+import { DocumentStorePlugin } from './backend/DocumentStorePlugin';
+import { ListenerInitializer } from './backend/ListenerInitializer';
+import { NoDocumentStorePlugin } from './backend/NoDocumentStorePlugin';
+import { NoQueryHandlerPlugin } from './backend/NoQueryHandlerPlugin';
+import { QueryHandlerInitializer } from './backend/QueryHandlerInitializer';
+import { QueryHandlerPlugin } from './backend/QueryHandlerPlugin';
+import { Subscribe } from './listener/Subscribe';
 
 dotenv.config();
 
-// Default to no backend plugin
-let loadedBackendPlugin: MeadowlarkBackendPlugin = NoMeadowlarkBackendPlugin;
+// Default to "no" plugin
+let loadedDocumentStore: DocumentStorePlugin = NoDocumentStorePlugin;
+let loadedQueryHandler: QueryHandlerPlugin = NoQueryHandlerPlugin;
+
+async function loadDocumentStore(pluginNpmName: string) {
+  return ((await import(pluginNpmName)) as DocumentStoreInitializer).initializeDocumentStore();
+}
 
-// Dynamic import
-async function loadBackendPlugin(pluginNpmName: string) {
-  return ((await import(pluginNpmName)) as any).initializeBackendPlugin();
+async function loadQueryHandler(pluginNpmName: string) {
+  return ((await import(pluginNpmName)) as QueryHandlerInitializer).initializeQueryHandler();
+}
+
+async function loadListener(pluginNpmName: string) {
+  return ((await import(pluginNpmName)) as ListenerInitializer).initializeListener(Subscribe);
 }
 
 // IIFE for top-level await
 (async () => {
-  if (process.env.BACKEND_PLUGIN_NAME) {
+  if (process.env.DOCUMENT_STORE_PLUGIN != null) {
+    try {
+      loadedDocumentStore = await loadDocumentStore(process.env.DOCUMENT_STORE_PLUGIN);
+    } catch (e) {
+      const message = e instanceof Error ? e.message : 'unknown';
+      Logger.error(
+        `Unable to load document store plugin "${process.env.DOCUMENT_STORE_PLUGIN}". Error was ${message}`,
+        null,
+      );
+      throw e;
+    }
+  }
+
+  if (process.env.QUERY_HANDLER_PLUGIN != null) {
+    try {
+      loadedQueryHandler = await loadQueryHandler(process.env.QUERY_HANDLER_PLUGIN);
+    } catch (e) {
+      const message = e instanceof Error ? e.message : 'unknown';
+      Logger.error(`Unable to load query handler plugin "${process.env.QUERY_HANDLER_PLUGIN}". Error was ${message}`, null);
+      throw e;
+    }
+  }
+
+  // If this works out, we'll discover listeners dynamically - no need to specify
+  if (process.env.LISTENER1_PLUGIN != null) {
```

Comment:
For now, hardcoded the possibility of one or two listener plugins via env variables. No limit in practice though.

Reply:
Seeing this, and having seen the code above, I now better understand. This isn't (for now) a "subscribe here" polling mechanism, although a listener plugin might provide that. The listener is in fact running inline with the main process. Which is very interesting and worthwhile.